File: Benchmarks/4.3-carga-graficos-complejos/prueba-2/javascript/worker.js
self.onmessage = async (e) => {
    if (e.data.type === "runBenchmark") {
        try {
            const { numSeries, numPoints } = e.data.payload;
            const startTime = performance.now();

            const dataGenStart = performance.now();
            const x = new Float64Array(numPoints);
            for (let i = 0; i < numPoints; i++) {
                x[i] = (10 * i) / (numPoints - 1);
            }

            const ys = [];
            for (let i = 0; i < numSeries; i++) {
                const y = new Float64Array(numPoints);
                for (let j = 0; j < numPoints; j++) {
                    y[j] = Math.sin(x[j] + i) + normalRandom(0, 0.1);
                }
                ys.push(y);
            }
            const dataGenTime = performance.now() - dataGenStart;

            const renderStart = performance.now();
            const traces = ys.map((y, i) => ({
                x: x,
                y: y,
                mode: "lines",
                name: `Serie ${i + 1}`,
                line: { width: 1 },
            }));

            const layout = {
                title: "Series Temporales (JavaScript)",
                width: 700,
                height: 500,
            };

            const renderTime = performance.now() - renderStart;
            const totalTime = performance.now() - startTime;

            const estimatedMemBytes =
                x.byteLength + ys.reduce((acc, y) => acc + y.byteLength, 0);
            const memoryUsageMB = (estimatedMemBytes / 1024 ** 2).toFixed(2);

            self.postMessage({
                type: "result",
                payload: {
                    metrics: {
                        dataGenTime: dataGenTime,
                        renderTime: renderTime,
                        totalTime: totalTime,
                        mem: memoryUsageMB,
                    },
                    graphData: { traces, layout },
                },
            });
        } catch (error) {
            self.postMessage({
                type: "error",
                payload: error.message,
            });
        }
    }
};

function normalRandom(mean = 0, stdev = 1) {
    const u = 1 - Math.random();
    const v = Math.random();
    const z = Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v);
    return z * stdev + mean;
}

self.postMessage({ type: "ready" });

File: Benchmarks/4.3-carga-graficos-complejos/prueba-2/python/worker.py
# python/worker.py
import time
import numpy as np
from pyscript import sync


def js_run_py_benchmark(num_series, num_points):
    start_time = time.perf_counter()

    x = np.linspace(0, 10, num_points)
    rng = np.random.default_rng()
    ys = [np.sin(x + i) + rng.normal(0, 0.1, num_points)
          for i in range(num_series)]
    data_gen_time = (time.perf_counter() - start_time) * 1000
    mem = (x.nbytes + sum(y.nbytes for y in ys)) / (1024**2)

    # Construimos datos planos
    traces = []
    for i, y in enumerate(ys):
        traces.append({
            "x": x.tolist(),
            "y": y.tolist(),
            "mode": "lines",
            "name": f"Serie {i+1}"
        })

    layout = {
        "title": "Series Temporales (PyScript)",
        "width": 700,
        "height": 500
    }

    total_time = (time.perf_counter() - start_time) * 1000

    return {
        "traces": traces,
        "layout": layout,
        "data_gen_time": data_gen_time,
        "render_time": total_time - data_gen_time,
        "total_time": total_time,
        "memory": mem
    }


sync.js_run_py_benchmark = js_run_py_benchmark
